### DISCUSSION TITLE
Pines arbitrarios para max7219

### DISCUSSION THREAD

**Discussion by Veroledez** [tttapa/Control-Surface#675] [Q&A]
Hola Peter.
Estoy usando una matriz rgb pwm con 74hc595 por lo que SPI esta por defecto para esta matriz, el problema radica en que al usar tambien SPI para el max7219 la matriz se detiene y pierde el pwm, son pequeños instantes pero visibles lamentablemente, asi que me preguntaba si es posible usar otros pines para manejar el max7219, con la libreria de MD_Parola y otras se puede, pero no pude introducir los leds en un vuleds tal vez por la forma como se declaran los pines.

No encontre ninguna libreria para max  arduinolibraries.info que llame a los pines como su implementacion por ejemplo.  " max7219.pin(63) ",  
hay forma de usar otros pines o cual seria su concejo.

Uso esp32.

Muchas gracias. Excelente trabajo.

**Comment by tttapa** [comment]
The ESP32 has a second SPI port you can use, see https://github.com/espressif/arduino-esp32/blob/master/libraries/SPI/examples/SPI_Multiple_Buses/SPI_Multiple_Buses.ino. You could use that one for the MAX7219. The latest version of Control Surface allows you to specify an arbitrary SPI interface and an arbitrary SPI driver for the MAX7219 class, see the `SPIDriver` template parameter and the `spi` constructor argument of https://tttapa.github.io/Control-Surface-doc/Doxygen/de/de9/classAH_1_1MAX7219.html.

If you want to use arbitrary pins, you could of course use bit-banged SPI: just create an SPI driver class with some stubs:
```cpp
(untested)

// Output only (uses shiftOut)
class SPIDriver {
  public:
    SPIDriver(uint8_t mosi, uint8_t clk) : mosi(mosi), clk(clk) {}

    void begin() {
        pinMode(mosi, OUTPUT);
        pinMode(clk, OUTPUT);
    }
    
    void beginTransaction(SPISettings) { /* blank, settings ignored */ }
    void endTransaction() { /* blank */ }

    void transfer(uint8_t data) {
        ::shiftOut(mosi, clk, MSBFIRST, data); // no input support here
    }

    uint8_t mosi, clk;
};

MAX7219<1, SPIDriver> max7219 {SPIDriver{MOSI, SCK}, SS};
```
